### PR TITLE
CSP-783 Implements pagination on network events search page

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    [TestCase(3, 5, 6, 8, 1, true, true)]
+    [TestCase(4, 5, 6, 8, 1, true, true)]
+    [TestCase(5, 5, 6, 8, 1, true, true)]
+    [TestCase(6, 5, 6, 7, 1, true, false)]
+    [TestCase(1, 5, 7, 7, 1, false, true)]
+    [TestCase(2, 5, 7, 8, 1, true, true)]
+    [TestCase(3, 5, 100, 8, 1, true, true)]
+    [TestCase(4, 5, 100, 8, 2, true, true)]
+    [TestCase(5, 5, 100, 8, 3, true, true)]
+    [TestCase(6, 5, 100, 8, 4, true, true)]
```
[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:37]: Some of these you may have already covered

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    public void PopulatesLinkItem(int currentPage, int pageSize, int totalPages, int totalLinkItems, int firstPageExpected, bool isPreviousExpected, bool isNextExpected)
+    {
+        var linkItems = Enumerable.Range(firstPageExpected, 6);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:40]: There is no assertion on firstExpectedPage, also I think there should be an assertion on lastExpectedPage

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+using SFA.DAS.ApprenticeAan.Domain.Constants;
+using SFA.DAS.ApprenticeAan.Web.Models.NetworkEvents;
+using SFA.DAS.ApprenticeAan.Web.Services;
+
+namespace SFA.DAS.ApprenticeAan.Web.UnitTests.Services;
+
+public class QueryStringParameterBuilderTests
+{
+
+    [Test, AutoData]
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Services/QueryStringParameterBuilderTests.cs:12]: Expected one test per field with two test cases one where value is not null and other where value is null

File: src/SFA.DAS.ApprenticeAan.Web/Services/FilterBuilder.cs [hunk 4]
```diff
@@ -32,6 +32,12 @@ public static List<SelectedFilter> Build(GetNetworkEventsRequest request, IUrlHe
         return filters;
     }
 
+    public static string BuildFullQueryString(GetNetworkEventsRequest request, IUrlHelper url)
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Services/FilterBuilder.cs:35]: This should already be in place?

Comment (DevenKShah) [reply src/SFA.DAS.ApprenticeAan.Web/Services/FilterBuilder.cs:35]: Also is this missing coverage?

File: src/SFA.DAS.ApprenticeAan.Web/Views/Onboarding/ShutterPage.cshtml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿@using SFA.DAS.ApprenticeAan.Web.Models.Onboarding
 @model ShutterPageViewModel
 @{
-    ViewData["Title"] = "Shutter Page";
+    ViewData["Title"] = "Shutter CurrentPage";
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Views/Onboarding/ShutterPage.cshtml:4]: yep, will fix them all

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs (hunk 2 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:40]: yes

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:37]: I will add all of these.  Some of them aren't right if we're stick to this: https://assessors.apprenticeships.education.gov.uk/find-an-assessment-opportunity/ShowApprovedStandardsPerPage?standardsPerPage=500#approved 
For example, when there are only 2 pages, it shows '1 2 Next >>'.  I will add all the cases above, but conforming the the approved standards pagination rules.

[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:37]: Ok I missed that, it just means that in that case the hasNext is true.

File: src/SFA.DAS.ApprenticeAan.Web/Views/NetworkEventDetails/CancellationConfirmation.cshtml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ﻿@using SFA.DAS.ApprenticeAan.Web.Models.NetworkEvents;
 
 @{
-    ViewData["Title"] = $"Confirmation Page — Cancel attendance";
+    ViewData["Title"] = $"Confirmation CurrentPage — Cancel attendance";
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Views/NetworkEventDetails/CancellationConfirmation.cshtml]: fixed

File: src/SFA.DAS.ApprenticeAan.Web/Views/NetworkEventDetails/SignUpConfirmation.cshtml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 ﻿@using SFA.DAS.ApprenticeAan.Web.Models.NetworkEvents;
 
 @{
-    ViewData["Title"] = $"Confirmation Page — Sign up event";
+    ViewData["Title"] = $"Confirmation CurrentPage — Sign up event";
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Views/NetworkEventDetails/SignUpConfirmation.cshtml]: fixed

File: src/SFA.DAS.ApprenticeAan.Web/Models/PaginationViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    public PaginationViewModel(int currentPage, int pageSize, int totalPages, string baseUrl)
+    {
+        CurrentPage = currentPage;
+        PageSize = pageSize;
+        TotalPages = totalPages;
+        BaseUrl = baseUrl;
+
+        if (totalPages == 0)
+            return;
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Models/PaginationViewModel.cs]: fixed

File: src/SFA.DAS.ApprenticeAan.Web/Models/PaginationViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+        var startPage = currentPage < 4 ? 1 : currentPage - 2;
+        if (totalPages <= 6)
+        {
+            startPage = 1;
+        }
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Models/PaginationViewModel.cs]: I like it.  Sonarcloud is reporting the ternary embedded logic as a code smell, but I think its coherent and fit for purpose.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        PaginationViewModel sut = new(currentPage, pageSize, totalPages, BaseUrl);
+
+        sut.LinkItems.Count.Should().Be(totalLinkItems);
+
+        sut.LinkItems.First(s => s.Text == currentPage.ToString()).HasLink.Should().BeFalse();
+        sut.LinkItems.Where(s => s.Text != currentPage.ToString()).All(s => s.HasLink).Should().BeTrue();
+
+        foreach (var text in linkItems)
+        {
+            //  sut.LinkItems.Exists(s => s.Text == text.ToString()).Should().BeTrue();
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs]: removed

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    [Test]
+    public void ReturnsNoLinksWhenNoTotalPages()
+    {
+        PaginationViewModel sut = new(1, 5, 0, BaseUrl);
+        sut.LinkItems.Count.Should().Be(0);
+    }
+
+    [TestCase(1, 5, 6, 7, 1, false, true)]
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs]: done

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs]: yes.  I've added all these.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Services/QueryStringParameterBuilderTests.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Services/QueryStringParameterBuilderTests.cs:12]: done

File: src/SFA.DAS.ApprenticeAan.Web/Services/FilterBuilder.cs (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Services/FilterBuilder.cs:35]: Coverage added. Not sure I understand your first comment.  I needed to have a public method to build the entire query string, to pass it into the pagination as the baseurl to add page/pagesize to, in each pagination link.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs]: basically the principle is if youre not on the first page, the '<< previous' shows, and if you're not on the last page, the 'next>>' shows

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/NetworkEventsControllerTests.cs [hunk 12]
```diff
@@ -70,14 +73,52 @@ public class NetworkEventsControllerTests
...
 
-        outerApiMock.Verify(o => o.GetCalendarEvents(It.IsAny<Guid>(), fromDateFormatted, toDateFormatted, eventFormats, eventTypes, regions, It.IsAny<CancellationToken>()), Times.Once);
+        outerApiMock.Verify(o => o.GetCalendarEvents(It.IsAny<Guid>(), It.IsAny<Dictionary<string, string[]>>(), It.IsAny<CancellationToken>()), Times.Once);
+    }
+    [Test, MoqAutoData]
+    public void GetCalendarEventsNoFilters_ReturnsApiResponse(
+   [Frozen] Mock<IOuterApiClient> outerApiMock,
+   [Greedy] NetworkEventsController sut,
+           GetCalendarEventsQueryResult expectedResult,
+           Guid apprenticeId)
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/NetworkEventsControllerTests.cs:91]: Indentation

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        if (isNextExpected)
+        {
+            sut.LinkItems.First(s => s.Text == PaginationViewModel.NextText).Url.Should().Be(BaseUrl + "?page=" + (currentPage + 1) + "&pageSize=" + pageSize);
+        }
+        else
+        {
+            sut.LinkItems.Exists(s => s.Text == PaginationViewModel.NextText).Should().BeFalse();
+        }
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Models/PaginationViewModelTests.cs:104]: Can these be merged with the if condition on line # 56

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**